provider/source/src/units/categorized_display_name.rs

```diff
@@ -0,0 +1,153 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use std::collections::HashSet;
+
+use crate::cldr_serde::{self};
+use crate::SourceDataProvider;
+
+use icu::experimental::dimension::provider::units::categorized_display_name::{
+    AreaDisplayNameV1, DurationDisplayNameV1, LengthDisplayNameV1, MassDisplayNameV1,
+    VolumeDisplayNameV1,
+};
+use icu::experimental::dimension::provider::units::display_name::UnitsDisplayName;
+use icu::plurals::PluralElements;
+use icu_provider::prelude::*;
+use icu_provider::DataMarkerAttributes;
+
+macro_rules! impl_categorized_display_name_data_provider {
+    ($category:expr, $display_name:ident) => {
+        impl DataProvider<$display_name> for SourceDataProvider {
+            fn load(&self, req: DataRequest) -> Result<DataResponse<$display_name>, DataError> {
+                self.check_req::<$display_name>(req)?;
+
+                // This category is based on the Marker
+                let category = $category;
+                let (length, unit) = req
+                    .id
+                    .marker_attributes
+                    .split_once('-')
+                    .ok_or_else(|| DataErrorKind::InvalidRequest.into_error())?;
```

Comment:
Nit: Always call `with_debug_context()` or `with_req()` or similar.

Suggestion: you can use `.and_then()` in order to reduce the number of error handlers you need to write.

Reply:
do0ne